proposals/interface-types/working-notes/scenarios/chooser.md

```diff
@@ -0,0 +1,680 @@
+
+>Note the `-x` suffix signals that these functions are part of the exporting
+>module.
+
+>Note, like the import and export adapters themselves, they will be inlined as
+>part of the adapter fusion process.
+
+The returned value from `nondeterministic_choice` is wrapped up as an `own`ed
+allocation and lifted to a `string`.
+
+>Note we do not need to `own` the string memory of the arguments because we have
+>asserted that both the arguments to `nondeterministic_choice` will be the
+>_last_ references to the strings. However, only one of the C++ strings will be
+>deallocated -- the other is returned to us. We _do_ need to `own` the return
+>result, however.
+
+In addition to the Interface Types memory management, because we are using a
+non-trivial C++ structure, we have to invoke the appropriate constructors,
+access functions and destructors of our `shared_ptr` structure.
+
+This is achieved through some additional core helper functions that are exported
+by the core wasm module, although they would not typically be exported as
+Interface Type functions:
+
+* `shared-builder-x` is used to create and initialize a `shared_ptr`
+  structure. It takes a pointer as an argument and returns a `shared_ptr` with
+  an initial reference count of 1.
+
+* `shared-release-x` is used to decrement the reference count of a `shared_ptr`;
+  and if that results in the reference count dropping to zero _both the
+  underlying resource and the `shared_ptr` structure itself are disposed of_.
+
+* `shared-get-x` is used to access the underlying resource of a `shared_ptr`
+  without affecting it's reference count.
+
+* `access-utf8-x` is used to access the memory pointer for a string's text and
+  its length.
+
+In practice there would likely be additional support functions -- such as a
+version of accessing a shared resource whilst incrementing reference
+count. However, we do not need them in this scenario.
+
+## Calling the chooser
+
+We shall assume that the import to `nondeterministic_choice` were as though it
+was from a core WebAssembly import whose signature is:
+
+```wasm
+(func (import "" "chooser_")
+  (param i32 i32)
+  (result i32))
+```
+
+The two strings are passed as memory addresses of structures -- from which the
+text of the string and its length can be ascertained. The structure itself is
+unspecified; we will use access functions -- such as `access-utf8` as needed.
+
+>Note that although _we_ believe that the returned value will be the same as one
+>of the arguments, the limitations of Interface Types mean that the returned
+>string will be a copy of one of the arguments. 
+
+The import adapter for `chooser_` has to lift the two argument `string`s and
+lower the return value:
+
+```wasm
+(@interface func $stralloc-i
+  (param $str string)
+  (result i32)           ;; return a shared ptr of the string
+  local.get $str
+  string.size
+  call-export "malloc-i"     ;; create string entity of right size
```

Comment:
I think you should `own` the result of `malloc-i` with a call to free, making the return type `(owned i32)`, and adding an `owned.release` right before returning in `chooser_`.  Otherwise any exception between here and the final return would leak this memory.